src/mlir/dialect/VeronaTypes.cc

```diff
@@ -510,4 +511,114 @@ namespace mlir::verona
         abort();
     }
   }
+
+  FieldOp lookupClassField(ClassOp classOp, StringRef name)
+  {
+    assert(classOp != nullptr);
+    for (Operation& op : *classOp.getBody())
+    {
+      if (verona::FieldOp fieldOp = dyn_cast<verona::FieldOp>(op);
+          fieldOp && fieldOp.name() == name)
+      {
+        return fieldOp;
+      }
+    }
+    return nullptr;
+  }
+
+  LogicalResult lookupFieldTypes(
+    Operation* op,
+    ArrayRef<Type> origins,
+    StringRef name,
+    SmallVectorImpl<Type>& readElements,
+    SmallVectorImpl<Type>& writeElements,
+    bool exhaustive)
+  {
+    bool found = false;
+    for (Type origin : origins)
+    {
+      auto [readType, writeType] = lookupFieldType(op, origin, name);
+      assert((readType == nullptr) == (writeType == nullptr));
+      if (readType != nullptr)
+      {
+        readElements.push_back(readType);
+        writeElements.push_back(writeType);
+        found = true;
+      }
+      else if (exhaustive)
```

Comment:
What's the semantics of exhaustive?

Reply:
It's described in the header file. It indicates whether we want the field to be present in all elements of `origins`, or only in at least one. I don't really like the name, but I couldn't come up with anything better.

Reply:
Right, I did read the header comment and that didn't clarify much (to me, personally). Perhaps I don't quite get what the semantics of the field write is to have a good understanding of why only the last one matching would make sense. Do you have some code example that could help?

Reply:
> only the last one

It's not the last, it's at *least* one

> I don't quite get what the semantics of the field write is

The flag does not control whether this is a read or a write, it controls whether we are looking up in a join or meet type.

> Do you have some code example that could help?

When looking up a field in `join<T, U>`, the origin could be *either* a T or a U, so the field is required to be present in both types. When lookin up in `meet<T, U>`, the origin is *both* a T and a U. Therefore the field being present in just one of them is enough for us.

The degenerate case, when `origins` is empty, works as well. An empty meet (ie. top), is a type that allows any value. It therefore gives us no information about the contents, hence lookups always fail.

An empty join (ie. bottom) is a bit trickier. It is a type that has no allowed value, therefore any operation on it is allowed (since the code is unreachable anyway).

---

Tbf I'm not set on the utility of merging these two cases into one function, so if it causes too much confusion I'm happy to separate it into two different functions.

Reply:
> I'm not set on the utility of merging these two cases into one function, so if it causes too much confusion I'm happy to separate it into two different functions.

Alright, I've split it and I think it makes things much clearer. Let me know what you think.

Reply:
I was going to suggest `any|all` flag, but split looks cleaner. Thanks!